### ISSUE TITLE
Optional arguments in object variable type definition

### ISSUE THREAD

**Issue by prehor** [hashicorp/terraform#19898]
### Current Terraform Version
```
Terraform v0.12.0-alpha4 (2c36829d3265661d8edbd5014de8090ea7e2a076)
```

### Proposal
I like the `object` variable type and it would be nice to be able to define optional arguments which can carry `null` value too, to use:
```hcl
variable "network_rules" {
  default = null
  type = object({
    bypass = optional(list(string))
    ip_rules = optional(list(string))
    virtual_network_subnet_ids = optional(list(string))
  })
}

resource "azurerm_storage_account" "sa" {
  name = random_string.name.result
  location = var.location
  resource_group_name = var.resource_group_name
  account_replication_type = var.account_replication_type
  account_tier = var.account_tier

  dynamic "network_rules" {
    for_each = var.network_rules == null ? [] : list(var.network_rules)

    content {
      bypass = network_rules.value.bypass
      ip_rules = network_rules.value.ip_rules
      virtual_network_subnet_ids = network_rules.value.virtual_network_subnet_ids
    }
  }
```
instead of:
```hcl
variable "network_rules" {
  default = null
  type = map(string)
}

resource "azurerm_storage_account" "sa" {
  name = random_string.name.result
  location = var.location
  resource_group_name = var.resource_group_name
  account_replication_type = var.account_replication_type
  account_tier = var.account_tier

  dynamic "network_rules" {
    for_each = var.network_rules == null ? [] : list(var.network_rules)

    content {
      bypass = lookup(network_rules, "bypass", null) == null ? null : split(",", lookup(network_rules, "bypass"))
      ip_rules = lookup(network_rules, "ip_rules", null) == null ? null : split(",", lookup(network_rules, "ip_rules"))
      virtual_network_subnet_ids = lookup(network_rules, "virtual_network_subnet_ids", null) == null ? null : split(",", lookup(network_rules, "virtual_network_subnet_ids"))
    }
  }
}
```

**Comment by abymsft** [comment]
I'd really like to have this option too. Any plans of prioritizing this enahancement ?

**Comment by rverma-nikiai** [comment]
is there a workaround for this? I found myself using type=any

**Comment by sudoforge** [comment]
@rverma-nikiai No, there is not currently a workaround for this.

The best thing we can do is define the optional argument as such (where `foobar` is the optional property):

*`modules/myfoo/main.tf`*

```hcl
variable "foo" {
  type = object({
    name   = string
    id     = number
    foobar = any
  })
}

resource "null_resource" {
  triggers = {
    some_name   = var.foo.name
    some_id     = var.foo.id
    some_foobar = var.foo.foobar
  }
}
```

... and then call it, explicitly setting the optional properties to `null`

*`main.tf`*

```hcl
module "potato" {
  source = "./modules/myfoo"
  foo = {
    name   = "something-cool-is-afoot"
    id     = 1234567890
    foobar = null
  }
}
```

This allows you to do normal stuff like set defaults using a ternary check, eg.:

*`modules/myfoo/main.tf`*

```
locals {
  foobar_default = var.foo.foobar == null ? "default value for foo.foobar" : null
}

...
```

**Comment by thefotios** [comment]
Along the same lines, it would be useful to be able to declare partial defaults for complex types

For instance, something like
```tf
variable "with_default_path" {
  type = object({
    id = string
    path = string
  })
 
  default = {
    path = "/"
  }
}
```

**Comment by mildwonkey** [comment]
Hi folks! This is a great feature request, and getting use-cases is absolutely helpful, but so you know it best to react to the original issue comment with 👍which we [can and do report on](https://github.com/hashicorp/terraform/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc) during prioritization.
Thank you :)

**Comment by solsglasses** [comment]
> Along the same lines, it would be useful to be able to declare partial defaults for complex types
> 
> For instance, something like
> 
> ```terraform
> variable "with_default_path" {
>   type = object({
>     id = string
>     path = string
>   })
>  
>   default = {
>     path = "/"
>   }
> }
> ```

Absolutely, this should be part of the original post as well

👍 Optional values
👍 Partial default values

**Comment by sudoforge** [comment]
@solsglasses - please [add reactions to comments, pull requets, or issues][0] instead of creating a new comment simply to show agreement with someone. Thanks for helping to keep the community awesome, and the notification feed of the maintainers free of fluff!

[0]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/

**Comment by thefotios** [comment]
I'm not sure of the internals, but this may not be that difficult. Since type and defaults are already well defined, the parser would just need to allow both of them and then merge the defaults with the provided variable before validation. 

My proposal mostly dealt with optional values inside the type definition, not the argument itself being optional. The might get a little trickier. Having the argument be optional is going to lead to a lot of lines like:
```
for_each = var.network_rules == null ? [] : list(var.network_rules)
```
It may make more sense for the `default` value to match the `type` definition. So a `list(object(...))` would return an empty list. This way you can let the parser handle the null check and you can be guaranteed to get the proper `type` for iteration. This is assuming that an empty `list` is valid for the `type` definition of `list(object(...))`.

It could reduce a bunch of boilerplate code if there was an `optional` flag in the variable definition.

**Comment by solsglasses** [comment]
This doc looks very promising: https://www.terraform.io/docs/configuration/attr-as-blocks.html#arbitrary-expressions-with-argument-syntax

> If you are writing a reusable module that allows callers to pass in a list of objects to assign to such an argument, you may wish to use the `merge` function to populate any attributes the user didn't explicitly set, in order to give the module user the effect of optional arguments:

...but I'm still trying to figure out how to make it work with a `list(object)`. Anything that I try ends with an error. I know I'm doing it wrong. I'm not totally comfortable with all the v12 `for`/`for_each`/conditional features yet. I'll report back with my progress.

Here's a simple `list(object)` for an example, with the attribute `totally_optional_env_vars` being present in one object, but not the other.

```
myvar = [
  {
    name = "foo"
    os   = "linux"
  },
  {
    name = "bar"
    os   = "linux"
    totally_optional_env_vars = [
      {
        name  = "HELLO",
        value = "world"
      },
      {
        name  = "FOO",
        value = "bar"
      }
    ]
  }
]
```

**Comment by atavakoliyext** [comment]
One more proposal to add to the mix:

```
object({
  <FIELD> = <TYPE | { [type = <TYPE>,][ default = ...] }>
})
```

I think would allow more readable complex object definitions in cases where defaults could appear at any level. For a contrived example:

```
variable "pipeline" {
  type = object({
    name = string

    pipeline_type = {
      type = string        // implied from default & can be omitted
      default = "STANDARD"
    }

    stages = {
      type = list(object({
        name      = string
        cmd       = string
        must_pass = { default = false }
        }
      }))

      default = [
        {
          name      = "build"
          cmd       = "make all"
          must_pass = true
        },
        {
          name = "deploy"
          cmd  = "make deploy"
        },
      ]
    }
  })

  default = {
    name = "standard_build_pipeline"
    // everything else is taken from the type defaults
  }
}
```

**Comment by morgante** [comment]
For what it's worth, even the minimal behavior (accepting objects with omitted values and setting the omitted values to null) would be a major usability improvement. Defaults could be implemented within modules themselves.

**Comment by binlab** [comment]
Very needed possibility for maps and objects types.

**Comment by sidewinder12s** [comment]
Being able to set a `default = null` type I think would allow modules to set object type constraints for input variables or resources that have optional attributes, while still being able to set some type/object constraints.

**Comment by atavakoliyext** [comment]
My personal vote is also to add support for default values, rather than support for marking some fields as optional.

Defaults imply optionality, so you get optionals for free by supporting defaults. OTOH, if we add `optional(...)` support, and then later add defaults, then we'll have two different ways of defining optionality (specifying a default vs specifying `optional(...)`, the latter being equivalent to specifying a `null` default), which I think would be less preferred.

**Comment by cadavre** [comment]
Definitely @atavakoliyext proposal is best.

@prehor thefotios proposal:
```
variable "with_default_path" {
  type = object({
    id = string
    path = string
  })
 
  default = {
    path = "/"
  }
}
```

Why we cannot use it:
```
variable "with_default_path" {
  type = list(object({
    id = string
    path = string
  }))
 
  default = [{
    path = "/"
  }] # ???
  # or how?
  # this doesn't look reasonable
}
```

**Comment by lemontreeran** [comment]
Actually there are workarounds for optional arguments in object variables. The tricky part is still on the default value of the variable and using local variables. Use the locals variable to verify the optional parameter in "network_rules". Put it into null if not existing. The other resources would be able to refer to those local variables.

variable "network_rules" {
  default = {
    bypass = null,
    ip_rules = null,
    virtual_network_subnet_ids = null
  }
  type = object({
    bypass = list(string)
    ip_rules = list(string)
    virtual_network_subnet_ids = list(string)
  })
}

locals {
    network_rules = lookup(var.network_rules, "bypass", "no_bypass") == "no_bypass" ? null : var.network_rules,
    ip_rules = lookup(var.ip_rules, "ip_rules", "no_ip_rules") == "no_ip_rules" ? null : var.network_rules,
    virtual_network_subnet_ids = lookup(var.virtual_network_subnet_ids, "virtual_network_subnet_ids", "no_virtual_network_subnet_ids") == "no_virtual_network_subnet_ids" ? null : var.network_rules
}

**Comment by grimm26** [comment]
@lemontreeran This does not work.  I cleaned up your example:
```
variable "network_rules" {
  default = {
    bypass                     = null,
    ip_rules                   = null,
    virtual_network_subnet_ids = null
  }
  type = object({
    bypass                     = list(string)
    ip_rules                   = list(string)
    virtual_network_subnet_ids = list(string)
  })
}

locals {
  network_rules              = lookup(var.network_rules, "bypass", "no_bypass") == "no_bypass" ? null : var.network_rules
  ip_rules                   = lookup(var.network_rules, "ip_rules", "no_ip_rules") == "no_ip_rules" ? null : var.network_rules
  virtual_network_subnet_ids = lookup(var.network_rules, "virtual_network_subnet_ids", "no_virtual_network_subnet_ids") == "no_virtual_network_subnet_ids" ? null : var.network_rules
}
```
And provided a terraform.tfvars like so:
```
network_rules = {
  bypass = ["nope"]
}
```

The result:
```
> $ terraform plan
Error: Invalid value for input variable

  on terraform.tfvars line 1:
   1: network_rules = {
   2:   bypass = ["nope"]
   3: }

The given value is not valid for variable "network_rules": attributes
"ip_rules" and "virtual_network_subnet_ids" are required.
```

You cannot have optional elements of a well-defined variable object.

**Comment by lemontreeran** [comment]
> @lemontreeran This does not work. I cleaned up your example:
> 
> ```
> variable "network_rules" {
>   default = {
>     bypass                     = null,
>     ip_rules                   = null,
>     virtual_network_subnet_ids = null
>   }
>   type = object({
>     bypass                     = list(string)
>     ip_rules                   = list(string)
>     virtual_network_subnet_ids = list(string)
>   })
> }
> 
> locals {
>   network_rules              = lookup(var.network_rules, "bypass", "no_bypass") == "no_bypass" ? null : var.network_rules
>   ip_rules                   = lookup(var.network_rules, "ip_rules", "no_ip_rules") == "no_ip_rules" ? null : var.network_rules
>   virtual_network_subnet_ids = lookup(var.network_rules, "virtual_network_subnet_ids", "no_virtual_network_subnet_ids") == "no_virtual_network_subnet_ids" ? null : var.network_rules
> }
> ```
> 
> And provided a terraform.tfvars like so:
> 
> ```
> network_rules = {
>   bypass = ["nope"]
> }
> ```
> 
> The result:
> 
> ```
> > $ terraform plan
> Error: Invalid value for input variable
> 
>   on terraform.tfvars line 1:
>    1: network_rules = {
>    2:   bypass = ["nope"]
>    3: }
> 
> The given value is not valid for variable "network_rules": attributes
> "ip_rules" and "virtual_network_subnet_ids" are required.
> ```
> 
> You cannot have optional elements of a well-defined variable object.

Try adding the null value for those vars. I know it is ugly......
```
network_rules = {
  bypass                                  = ["nope"],
  ip_rules                                 = null,
 virtual_network_subnet_ids = null
}
```

**Comment by grimm26** [comment]
> Try adding the null value for those vars. I know it is ugly......
> 
> ```
> network_rules = {
>   bypass                      = ["nope"],
>   ip_rules                    = null,
>   virtual_network_subnet_ids  = null
> }
> ```

That means those attributes  are *not optional* if you have to supply a value, which is the whole point of this issue.

**Comment by sudoforge** [comment]
@grimm26 Yes, that is why the issue is still open. As @lemontreeran noted, and as I wrote [earlier in the thread](https://github.com/hashicorp/terraform/issues/19898#issuecomment-506108063), providing `null` values for the "optional" parameters is the only way to circumvent the issue.

**Comment by igorbrites** [comment]
Or following the first post, I'd like to pass default values like this:

```hcl
variable "network_rules" {
  default = null
  type = object({
    bypass                     = optional(list(string), ["teste"])
    ip_rules                   = optional(list(string), null)
    virtual_network_subnet_ids = optional(list(string), [])
    enabled                    = optional(bool, true)
  })
}
```

The `optional` could work like the `lookup` function (without the first `map` parameter, off course), receiving the type on the first parameter and the default value on the second one.

**Comment by mutt13y** [comment]
Where we have a map or list of map the solution should include the capability for the map key to be absent.

```
type = list(object({
  foo = optional(string,"foo_default")
  bar = optional(string)
}))
```
so if bar is not provided by the caller the key would be completely absent.

this way you can use the absence of the key to infer some more complex default action rather than having to use a rouge value for default.

**Comment by igtsekov** [comment]
Any idea if/when this will be addresed?

**Comment by heldersepu** [comment]
This is the most upvoted issue, can we get an update, please... 
@mildwonkey, @apparentlymart

**Comment by madsonic** [comment]
Without optional param in objects, it really limits the level of abstraction that can be designed in a module.

**Comment by yves-vogl** [comment]
I'm using the *null* pattern in lists with objects for dynamic blocks for CloudFront origins. And optional parameter would be nice to skip defining the http_port attribute when using https-only protocol policy. But as thinking about this you will fast come to the thought of „conditional optional parameters“ (e.g. http_port is only optional when corresponding protocol_policy parameters are set). This is imho against the declarative approach…

**Comment by RTodorov** [comment]
Wow, more than a year and 410 upvotes in the original post. @mildwonkey is there any chance this is gonna land in the roadmap at some point?

**Comment by solarmosaic-kflorence** [comment]
It is unfortunately easiest to work around this right now by using `type = map(any)` and `lookup(object, "key", "default")`, so long as you can put up with the terrible feeling of guilt it gives you.

**Comment by agitelzon** [comment]
I'd also like for this feature to be implemented. Here is what I did to get around not having optional arguments using merge and a bunch of for loops for AWS.

I can now use the variable var.websites in multiple resources to make sure the dns, SSL certificate, and cloudfront config all get created off of one variable while still allowing me the flexibility to have each website have slightly different parameters.

```
variable "websites" {
  type = any
}

websites = {
  "site1.example.com" = {
    enable_cf = true,
  }

  "site2.example.com" = {
    enable_cf = true,
  }
}

locals {
  default_cf_settings = {
    aliases    = null,
    enable_cf = false,
  }

  cf_settings_merged = {
    for key, value in var.websites : key => merge(local.default_cf_settings, value)
  }

  cf_settings = {
    for key, value in local.cf_settings_merged :
    key => value
    if value.enable_cf == true
  }
}

resource "aws_cloudfront_distribution" "cf_config" {
  for_each = local.cf_settings
  comment  = each.key
  aliases = each.value.aliases != null ? each.value.aliases : (length(regexall("\\.", each.key)) <= 1 ?
  ["www.${replace(each.key, "/.*?([\\w]+\\.[\\w]+)$/", "$1")}"] : ["${each.key}"])

  enabled         = true

{...everything else...}
}
```

**Comment by spirius** [comment]
Something like this can be used as a replacement. Supports required, optional, nested/complex structures and type-checking. Not very readable though.

```hcl
variable "config" {
  type = any
}

locals {
  config = {
    required = tostring(var.config.required)

    optional1 = tostring(try(var.config.optional1, "default1"))
    optional2 = [for v in try(var.config.optional2, ["default2", "default3"]): tostring(v)]

    sub = tomap(merge(
      {
        optional4 = "default4"
      },
      try(var.config.sub, {})
    ))
  }
}

output "config" {
  value = local.config
}
```

with input
```hcl
config = {
  required = "xxx"
  optional1 = "yyy"
  sub = {
    optional4 = "zzz"
  }
}
```

**Comment by hakro** [comment]
This feature would be great indeed.
I my case, I'm building a "Site" object that should have some default attributes. But since it's not possible to do it, I need to inject some default values when I call the module.

Example : 
`variable environments {
  description = "List of the environments attached to this site"
  type = map(object({
    ecs_cluster_name = string
    primary_url = string
    additional_urls = list(string)
  }))
}
`

When I call my module, I have to set `additional_urls = []`
Which works but is quite ugly, because most of my module invocations will have this empty value.
Ideally, it should be possible to set a default value for addition_urls, just like for other flat variables using the default keyword.

Cheers

**Comment by CarlosDomingues** [comment]
In addition to what others have said, I believe that in the end of day what most people want is a way of defining complex blocks the same way native resources do. For example, in `aws_instance` one can define a `network_interface` block:

```Terraform
  network_interface {
    network_interface_id = "${aws_network_interface.foo.id}"
    device_index         = 0
  }
```
[Reading the docs](https://www.terraform.io/docs/providers/aws/r/instance.html#network-interfaces) for this block and others, we can see that:

* Some attributes are optional and some are required.
* Some attributes have default, others do not.
* All attributes have description.
* The block itself has a description.
* The block itself can be optional.
* Attributes also can be complex blocks.

I'd like to be able to define similar interfaces in my modules.

**Comment by MarkKharitonov** [comment]
My scenario is described here - https://stackoverflow.com/questions/62105031/how-to-extend-terraform-module-input-variable-schema-without-breaking-existing-c?noredirect=1#comment109844759_62105031

Even without explicit default value, i.e. using fixed default values, like:
 - `false` for `bool`
 - `""` for `string`
 - 0 for number
 - [] for list
 - {} for map
This would be tremendously useful.

I do not really get it. This issue is open for 1.5 years. It is such a useful feature. It must be very difficult to implement, otherwise it would have been already. Can any one from the dev team provide some kind of a reason why this is so difficult to implement?

**Comment by LukeCarrier** [comment]
> I do not really get it. This issue is open for 1.5 years. It is such a useful feature. It must be very difficult to implement, otherwise it would have been already. Can any one from the dev team provide some kind of a reason why this is so difficult to implement?

Terraform is open source. Any one of us here could implement this feature if it were so simple.

HashiCorp are a private company with their own commercial goals and pressures. They’re releasing OSS for those of us who can’t afford their paid offerings for little commercial gain; they aren’t obliged to do this, let alone take on our feedback around what should be prioritised.

**Comment by heldersepu** [comment]
@MarkKharitonov here is something I got from the product managers:

> I’m saddened that this feature has left our community frustrated. Daniel and I only recently began with the team and have been working through the backlog of issues and concerns.  

> For 0.13 our goal was to tackle module expansion (count, for_each and depends_on) and deliver a significant update without the pain that many associated with our 0.12 release. The module-expansion work, when all related feature requests were combined, outweighed any other feature in terms of demand. We also felt this work delivered the broadest value to our community, which includes many who do not contribute or participate in GitHub. 

> By that reasoning however, #19898 is a strong possibility in our next major release cycle.

> One of the highest level goals we’ll be working on, over the next year, is increasing our communication with the community about what we’re doing and not doing. We will also engage in targeted discussions about what would be of most value. Sometimes we’ll do that in GitHub issues, and other times we’ll engage elsewhere. 
This turnaround won’t be overnight, but we are already hard at work.

`Petros Kolyvas <petros@hashicorp.com>`
`Daniel Dreier <ddreier@hashicorp.com>`

**Comment by MarkKharitonov** [comment]
@heldersepu - Thanks.

**Comment by jakshaym1234** [comment]
I'd really like to have this option too. Any plans of prioritizing this enhancement ?

**Comment by Arlington1985** [comment]
There is some workaround. See this blog:
https://binx.io/blog/2020/01/02/module-parameter-defaults-with-the-terraform-object-type/

**Comment by pirfalt** [comment]
Im very new to using terraform and this is one of the first things I ran into.

Since I have not used it before I may be missing something obvious, but is there any reason why we can't have named types? (Which could have there own defaults)

The first thing I attempted was something along the lines of:
```terraform
type "path_type" {
  type = string
  default = "/"
}

variable "with_default_path" {
  type = object({
    id = string
    path = path_type
  })
}
```

(Like I said, Im new to this so ignore the syntax part and focus on the "named custom types" which can have defaults part.)

**Comment by aaronsteers** [comment]
With 0.13 announced, and with variable validation moving forward - I'm wondering if the near-to-mid-term "best practice" should be:

1. declare these complex types as maps instead of declaring them as objects.
2. use validation rules to ensure:
    a. required keys are present
    b. any keys provided which are neither required nor optional would raise "an error indicating unknown key"

In absense of support for "optional" in the nearterm, does the above give us the desired behavior or is there anything additional I am missing from functional requirements?

**Comment by ritzz32** [comment]
> With 0.13 announced, and with [custom variable validation](https://www.terraform.io/docs/configuration/variables.html#custom-validation-rules) moving forward - I'm wondering if the near-to-mid-term "best practice" should be:
> 
> 1. declare these complex types as maps instead of declaring them as objects.
> 2. use validation rules to ensure:
>    a. required keys are present
>    b. any keys provided which are neither required nor optional would raise "an error indicating unknown key"
> 
> In absense of support for "optional" in the nearterm, does the above give us the desired behavior or is there anything additional I am missing from functional requirements?

The problem with using a map instead of a well define object is with a map all of the values have to be the same type.

For example if I declare the type as map(string) all of the map values have be to strings. If I declare the type as map or map(any). The values can be of any type, but all of the values in the map have to be the same type. Where a well defined object I can have sting keys, but each key may be a different type.

**Comment by aaronsteers** [comment]
@ritzz32 - Got it - and yes, I think you are right. My own use case also requires that some of the object's properties be strings (for instance) and some be lists of strings. So I think we are still blocked, as you say:

1:
> The problem with using a map instead of a well define object is with a map all of the values have to be the same type.

2:
> If I declare the type as map or map(any). The values can be of any type, but all of the values in the map have to be the same type.

**Comment by tomasbackman** [comment]
Im currently doing more or less as @aaronsteers suggested above with variable validation (`try()`), but have set `type =  any` directly for my complex object variables.. so not even declaring it as a map and thus getting around the "all having to be same type problem" @ritzz32  brought up..  it works for now, but does not feel very good..

**Comment by aaronsteers** [comment]
@tomasbackman - thanks very much for this info. So, basically we don't set it up as a dict - and it'll be interpreted perhaps as an object internally (of parent class 'any'). Very clever!

I have been struck by another related issue which I think should be combined with this one. Specifically, the problem with checking the presence of a nullable object.

Take this example:

```
variable "s3_logging_bucket" {
  description = "Optional. The bucket where logs should be written. If omitted, logs will not be preserved"
  type            = string
  default        = null
}

resource "s3_bucket_policy" "my_bucket_policy" {
  count = var.s3_logging_bucket == null ? 0 : 1
  // ...
}
```

Seems straightforward enough. It works fine if you pass `null`, it works fine if you pass nothing at all, and it works fine if you pass a string. However, if you pass in the output of another module, for instance if you pass `s3_logging_bucket = module.data_lake.logging_bucket` it will fail because during static analysis it cannot detect what `count` will be ahead of time and you'll get the error that looks like [value of `count` cannot be computed](https://docs.cloudposse.com/troubleshooting/terraform-value-of-count-cannot-be-computed/).

This error makes sense at first - terraform needs to deterministically know how many resources it's creating/tracking - except until you realize that the module is never going to output a non-null bucket name, and then this becomes a very frustrating issue with no good solution. The only solution I've found is to add another boolean value like `save_logging_to_s3` that is reduntant, but deterministic.

Now, tying it back together, this pattern could be solved in the same way that we can solve for optional arguments in object variable type definitions: that is, to introduce `optional(str)` and `required(str)` - with a smart default to one or the other depending upon context.

In my `count` example above, we solve it simply by declaring the output of the `data_lake` like so: `output "s3_logging_bucket { type = required(str) }`. This is then a promise to any consumer of the module outputs that `module.data_lake.s3_logging_bucket == null` always will result in `false`. It can then be calculated deterministically and our example `count` calculation will know that if it's getting its variable from that output, it's always going to be not null - even though we won't know the actually text value until the resource is deployed to AWS and AWS tells us the actual value.

The added benefit of this approach is that it solves other patterns as well. Currently, for any required module input, you can force the user to provide a value (by just not providing a default one), but you can't enforce that the user doesn't just pass `null` instead of an actual value. But in the new world, it would be simple: if you need the input to be non-null, you just specify `type = required(str)` and anyone who tries to pass null as an input will get a simple-to-debug error messages without any further headache.

**Comment by aaronsteers** [comment]
Fwiw, my proposal above is also very similar to the proposal noted above from @mutt13y - except I omit the default missing behavior in the second argument, opting for a syntax that is more of a type wrapper, is generic across non-object-related topics, and add on a proposal for `required()` which is the inverse of `optional()`. 

> Where we have a map or list of map the solution should include the capability for the map key to be absent.
> 
> ```
> type = list(object({
>   foo = optional(string,"foo_default")
>   bar = optional(string)
> }))
> ```
> 
> so if bar is not provided by the caller the key would be completely absent.
> 
> this way you can use the absence of the key to infer some more complex default action rather than having to use a rouge value for default.

**Comment by wdec** [comment]
Allow me to add a point that seems to be missing from all the proposals above and which also relates to what appears to be a bit of a kludge in how objects are defined. It's perhaps time to step back and re-look at the object definition per so.
If object member fields would follow the same definition syntax as variables, then pretty much no additional syntax rules would be required and it would be neater and more intuitive than the current form of object declaration. It would also allow validation syntax to apply to the object fields, which today require quite a lot of acrobatics to achive. 

In example form, it would be great to define objects including nullable optional fields like so:

```
variable "VMCluster" {
  type = object({
    variable "VMSize"  {
      type = string
      default = "Small"
      validation {
        condition = can(regex("^Small$|^Medium$|^Large$", var.VMSize))
        error_message = "Invalid VM Size."
    }
    variable "OptionalTag" {
      type = string
      default = null
}
```

**Comment by michelzanini** [comment]
I would like to suggest a simplified version of @wdec suggestion.

This what we can do today:

```
variable "my_cluster" {
  type = object({
    size = string,
    tag = string
  })
}
```

The idea is support an alternative syntax that supports the same options as the root variable:

```
variable "my_cluster" {
  type = object({
    size = {
      type = string
      default = "Small"
      validation {
        condition = can(regex("^Small$|^Medium$|^Large$", var.size))
        error_message = "Invalid VM Size."
      }
   }
   tag = {
      type = string
      default = null
   }
})
}
```

Advantage is that looks familiar as its the same attributes supported by the root variable.
It improves a bit @wdec suggestion by removing the need to write `variable` inside which is a bit confusing since the nested type is not really a variable on it's own.

This syntax improves a lot on using `any` as it can improve code that uses the `lookup` function. It also adds type validation with clear error messages. Another nice thing is to be able to do validation on inner properties.

Thanks.

**Comment by michelzanini** [comment]
Same suggestion seems to be described on this issue https://github.com/hashicorp/terraform/issues/24810

**Comment by michelzanini** [comment]
By the way, if you sort issues by reactions, this is issue comes up at the top over more than one thousand issues:

https://github.com/hashicorp/terraform/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc+

So please maintainers, consider it for the next version of Terraform, please!

**Comment by atavakoliyext** [comment]
> Same suggestion seems to be described on this issue #24810

This is also roughly what #22449 suggests (supporting defaults for objects with the same semantics as variables, where existence of a default implies optionality)

**Comment by CarlosDomingues** [comment]
Lots of cool suggestions. Not sure if that's the right place to mention it, but I also believe that a robust type system should support an `either` type or something similar to allow mutually exclusive arguments. [Example](https://www.terraform.io/docs/providers/aws/r/launch_template.html):

> `name` - The name of the launch template. If you leave this blank, Terraform will auto-generate a unique name.
>`name_prefix` - Creates a unique name beginning with the specified prefix. Conflicts with `name`.

Also:

> `encrypted` - Enables EBS encryption on the volume (Default: false). Cannot be used with `snapshot_id`.

There's also the case in which for an argument to be specified, another argument should have a specific value. [Example](https://www.terraform.io/docs/providers/aws/r/launch_template.html):

> `kms_key_id` - The ARN of the AWS Key Management Service (AWS KMS) customer master key (CMK) to use when creating the encrypted volume. `encrypted` must be set to `true` when this is set.

**Comment by michelzanini** [comment]
When writting Go code for a Terraform provider it's possible to define a complex schema such as:

```
"single_header": {
				Type:     schema.TypeList,
				Optional: true,
				MaxItems: 1,
				Elem: &schema.Resource{
					Schema: map[string]*schema.Schema{
						"name": {
							Type:     schema.TypeString,
							Required: true,
							ValidateFunc: validation.All(
								validation.StringLenBetween(1, 40),
								// The value is returned in lower case by the API.
								// Trying to solve it with StateFunc and/or DiffSuppressFunc resulted in hash problem of the rule field or didn't work.
								validation.StringMatch(regexp.MustCompile(`^[a-z0-9-_]+$`), "must contain only lowercase alphanumeric characters, underscores, and hyphens"),
							),
						},
					},
				},
			},

```

We just want to have the same on Terraform modules.

**Comment by robbruce** [comment]
TypeScript has a good way of declaring a property on an interface as optional

```typescript
interface MyObject {
  id:    string;
  name?: string;
}
```

The `?` make the `name` property optional. Maybe simple way to declare this for HCL could be

```hcl
object({
  id    = string
  name? = string
})
```

**Comment by aaronsteers** [comment]
What about adding support for "configuration blocks" in modules just like we can do for resources.

In a resouce, I can do this:

```hcl
resource "aws_security_group" "allow_tls" {
  name        = "allow_tls"
  vpc_id      = aws_vpc.main.id

  ingress {
    description = "TLS from VPC"
    from_port   = 443
    to_port     = 443
    protocol    = "tcp"
    cidr_blocks = [aws_vpc.main.cidr_block]
  }

  ingress {
    from_port   = 22
    to_port     = 22
    protocol    = "ssh"
    cidr_blocks = [aws_vpc.main.cidr_block]
  }

  egress {
    from_port   = 0
    to_port     = 0
    protocol    = "-1"
    cidr_blocks = ["0.0.0.0/0"]
  }
}
```

And each block can optionally be specified multiple times, each can have optional params, and each on the backend can interact with or leverage the defaults of the underlying config.

Is there any plans to move modules in this direction as well?

(Returning again to this issue after a while, as it is becoming another development roadblock as I migrate everything to TF 0.13 and as we can now have modules-calling-a-dynamic-count-of-modules.)

**Comment by aaronsteers** [comment]
Please watch and vote for this related issue if it would resolve your use case: https://github.com/hashicorp/terraform/issues/26051

In short:

```
variable "ingress" {
  type = block

  # Nested variables okay within 'block' variable type
  variable "from_port" {
    type = int
  }
  # Nested variables okay within 'block' variable type
  variable "to_port" {
    type = int
  }
  variable "description" {
    type    = string
    default = null # optional variables work just as they do today
  }
}
```

Which would looks exactly like configuration blocks that are already supported in today for resources (but not for modules):

```hcl
module "mymodule" {
  source = "..."
  ingress {
     from_port = 880
     to_port =  880
  }
  ingress {
     description = "Totally optional."
     from_port = 22
     to_port =  22
  }
}
```

Terraform would still output a list of objects, but because terraform is parsing the defaults block-by-block, each value in can be optional while still having the deterministic schema needed by objects today.

Inspired in part by @wdec and @michelzanini 's comments above.

**Comment by dylanturn** [comment]
Could something like this work?
```
variable "helm_parameters" {
  type        = list(object({
    name: string,
    value: any,
    force_string: bool,
  },{
    name: "default_name_string_or_null",
    value: "default_name_string_or_null",
    force_string: false
  }))
  description = "Parameters that will override helm_values"
  default     = []
}
```

**Comment by luanhdeandrade** [comment]
Will this still be addressed ?

**Comment by ocervell** [comment]
+1 it's been more than 2 years, this prevents creating complex variable types with Terraform (e.g: list of maps with optional fields, or maps with optional fields and more than 1 type within the map ...)

**Comment by jefftucker** [comment]
I've found a workaround for this, tested in v0.12.x but I don't see why it wouldn't work for v0.11.x also and probably v0.13x.  This is EXTREMELY ugly depending on the complexity of your input, but it works and I'm using it so that I can specify only the data I need in an input variable somewhere instead of having to fill out every last field.  It works for nested objects as well, but it gets uglier for each layer.  The trick is that Terraform is happy to accept "any" as a type, and then the merge() function works both with maps and with objects.  Imagine I have a module called "account" that has an input called "users" defined as such:

```
# account/variables.tf
variable "users" {
     type = map(object({
                email = string, # I want this to be required and get an error if it's missing
                readonly = bool, # I want this to default to false
                enabled = bool # I want this to default to true
               }))
}
```

Now for each user, I don't want to specify anything besides the email (the key will be their username) unless I need to, so my input in my inputs.auto.tfvars file would look like this:

```
users = { jsnuffy = {email = "jsnuffy@example.com},
                joeyjojo = {email = "joeyjojojr@example.com", readonly = true}
} 
```

Ok, now here's the fun part.  I need to "fix up" the missing stuff with some defaults from somewhere so that the module gets the correct types and doesn't complain and/or crash Terraform (I've found a few dozen ways to crash it in working on this).  My root module looks like this:

```
# main.tf
variable "users" {
    type = any,  # this is really critical or it'll try to coerce your input into something that you don't want
   default = {}  # this is optional; require the value or don't.
}

# hold my beer
locals {
    # where's the default for the input?  Let's just put it here!  
    defaultUser = {
         readonly = false,
         enabled = true
    }

    # THE ORDER HERE MATTERS!  the default comes first in the merge() function
    actual_users = { for k, v in var.users:                       
                        k => merge(local.defaultUser, v)
                  }
}

# now, local.actual_users has the right fields since if they were missing, the defaultUser variable had them and anything
# overwritten in whatever happens to be in the var.users input will be overwritten from the merge() function here.
# This is what I can pass to the module:

module "accounts" {
    src = "path/to/accounts"
    # this has been fixed up, so everything will match the type 
    # UNLESS the email is missing since we didn't include that with the default object that we passed to merge()
    users = local.actual_users 
                                              
}
```

That's all there is to it.  If you have a nested value, you do the same thing but introduce two more intermediary variables.  The first has the defaults for everything that is not nested.  The second takes the keys for all the inputs and declares ONLY the nested module, then does another loop calling merge() on the key/values in the nested input (use lookup() if you want that itself to be optional).  Now you have an intermediary local with all the fields that aren't nested, another with only the nested field, and then you can call merge() on these two locals to get your final local to pass in to the module.  You could probably abstract the locals by using a data-only module to fix up the inputs and make this slightly more clean, but the basic idea is to declare the input you actually want and then declare the "default" values elsewhere, then merge the input you get with the default values to construct a valid input that is then passed in to whatever you need to use.  Hope this helps someone; it took me a long time to find a way to make this work.

**Comment by itsankoff** [comment]
Guys I just found a working workaround for having great control over variables and their default values for composite type for Terraform 0.13.x. The solution uses the new variable validation approach. The workaround fixes 2 major limitations of Terraform:

1. Having a list of objects with different structure and still having type check
2. Having a way to specify optional properties in the list's objects

Here is a brief description:

Before we start with the code, here is some context. I want to define ingress rules for an AWS security_group module. The internal `aws_security_group` resource has 3 options for the source of the ingress rules of a security group - `cidr_blocks = list(string)`, `source_security_group_id = string` and `self = bool`. Now I want to have a unified structure on how to supply all the ingress rules from the module. It should look something like this:

```
module "security_group" "test" {
    source = "/path/to/module"

    vpc_id = "<vpc_id>"
    name = "test"

    ingress = [
        {
            protocol = "tcp"
            from_port = 0
            to_port = 65535
            description = "test sec group"
            source = {
                cidr_blocks = ["0.0.0.0/0"]
            }
        },
        {
            protocol = "tcp"
            from_port = 0
            to_port = 65535
            description = "test sec group"
            source = {
                source_security_group_id = "sg-<id>"
            }
        },
        {
            protocol = "tcp"
            from_port = 0
            to_port = 65535
            description = "test sec group"
            source = {
                 self = true
            }
        }
    ]
}
```

Now if I tried to define the input variable of the module like this:
```
variable "ingress" {
    type = string(object{
        protocol = string
        from_port = number
        to_port = number
        description = string
        source = any
     })
}
```

In theory, this should work. BUTTT it doesn't. The Terraform requires all objects within a list to look exactly the same. So, `source = any` gives us the flexibility, but the list limits us to have either all elements of the list to be with `source = { cidr_blocks = [...] }` or `source = { self = true|false }` and the same for `source_security_group_id`.

Can we do any better?

To bypass the list limitation we can use `type = any` for the whole input variable. But then we will lose the type checks during `plan` time which can result in misconfigurations during `apply` time.

So, can we do any better?

HOLD MY BEER!
```
# variables.tf
# This is a variable for a list of ingress rules for an AWS security_group module
variable "ingress" {
    type = any # we need this because Terraform requires objects to be absolutely equal for lists
    default = []    

    # we will use the new validation feature in 0.13 to do type check during plan time
    validation = {
        # The ternary operator returns true if the list is empty, because we want to support egress only security groups
        condition = length(var.ingress) > 0 ? length([
            for x in var.ingress : x
            if lookup(x, "protocol", null) != null &&
            if lookup(x, "from_port", null) != null &&
            if lookup(x, "source", null) != null        
        ]) == length(var.ingress) : true
        error_message = "Ingress rules validation error!"
    }
}
```

So let's see how this works. First, we define the type of `ingress` variable to be `any`. This removes the list limitation of object structure equality. But we want to have a loose type of validation. So what we do:
1. If the list is empty we return true so the validation can pass
2. If the list is not empty, then we filter all the objects that have all the required fields available. If there are rules that do not have any of the required fields (e.g. "protocol") then they will be missing from the result list.
3. In the end we check whether the filtered list has the same length as the provided list. If the length is not equal this means that we filtered an invalid rule from the list

A side effect of this approach is that we can still have type validation, but validate only the required fields and omit validation of optional fields like the description of the rules in our case. As you can see we don't `lookup` for the `description` field which means that we can provide it optionally.

I hope this can be of help until Terraform provides a more robust mechanism for type definition and type reuse.

cc @jefftucker

**Comment by jValdron** [comment]
I don't have time to try it myself, but instead of trying workarounds and all, but you guys should just try out v0.14.0-rc1 instead of trying to find a workaround. They've released this new experimental feature:

> `module_variable_optional_attrs`: When declaring an input variable for a module whose type constraint (type argument) contains an object type constraint, the type expressions for the attributes can be annotated with the experimental optional(...) modifier.
Marking an attribute as "optional" changes the type conversion behavior for that type constraint so that if the given value is a map or object that has no attribute of that name then Terraform will silently give that attribute the value null, rather than returning an error saying that it is required. The resulting value still conforms to the type constraint in that the attribute is considered to be present, but references to it in the recieving module will find a null value and can act on that accordingly.

**Comment by itsankoff** [comment]
This sounds interesting, I will check it out. But still, the list limitation for object structural equality is not resolved

**Comment by lperdereau** [comment]
This is really good. I think optional function should take a second parameter for default value.

For exemple I have this variable : 
```terraform
variable "ports" {
  type = list(object({
    name = string
    network_id = string
    subnet_id = string
    admin_state_up = optional(bool)
    security_group_ids = optional(list(string))
    ip_address = optional(string)
  }))
  description = <<EOF
The ports list
EOF
  default = []
}
```

For a list object it could be interesting to have a default value for optional attributes without a locals and defaults function like [here](https://github.com/hashicorp/terraform/blob/21195106691b694a07cb950cbd22dbe2c48f8c47/website/docs/configuration/functions/defaults.html.md)

Something like this:


```terraform
variable "ports" {
  type = list(object({
    name = string
    network_id = string
    subnet_id = string
    admin_state_up = optional(bool, true)
    security_group_ids = optional(list(string), ["general", "ssh"])
    ip_address = optional(string, null)
  }))
  description = <<EOF
The ports list
EOF
  default = []
}
```

**Comment by thiagolsfortunato** [comment]
I found this documentation with the experimental function `defaults` being used with `optional`. 

https://www.terraform.io/docs/configuration/functions/defaults.html

**Comment by ArchiFleKs** [comment]
I just tested this, I have an issue with optional values being set to null, is there a way to get them omited ?

My use case is this, on a majority of my module I use this pattern :
```
  aws-load-balancer-controller = merge(
    local.helm_defaults,
    {
      name                      = "aws-load-balancer-controller"
      namespace                 = "aws-load-balancer-controller"
      chart                     = "aws-load-balancer-controller"
      repository                = "https://aws.github.io/eks-charts"
      service_account_name      = "aws-load-balancer-controller"
      create_iam_resources_irsa = true
      enabled                   = false
      chart_version             = "1.0.5"
      version                   = "v2.0.0"
      iam_policy_override       = null
      default_network_policy    = true
      allowed_cidrs             = ["0.0.0.0/0"]
    },
    var.aws-load-balancer-controller
```

For now my variable `aws-load-balancer-controller` is type `any` with default to `{}`. This allow the module to work with just:

```
aws-load-balancer-controller = {
  enabled = true
}
```

Without further customization, if I set a validation like that:

```
variable "aws-ebs-csi-driver" {                                                      
+   description = "Customize aws-ebs-csi-driver helm chart, see `aws-ebs-csi-driver.tf`"    
+   type = object({                                                                    
+     atomic                = optional(bool)                                           
+     cleanup_on_fail       = optional(bool)                                           
+     dependency_update     = optional(bool)                                           
+     disable_crd_hooks     = optional(bool)                                           
+     disable_webhooks      = optional(bool)                                           
+     force_update          = optional(bool)                                           
+     recreate_pods         = optional(bool)                                           
+     render_subchart_notes = optional(bool)                                           
+     replace               = optional(bool)                                           
+     reset_values          = optional(bool)                                           
+     reuse_values          = optional(bool)                                           
+     skip_crds             = optional(bool)                                           
+     timeout               = optional(number)                                         
+     verify                = optional(bool)                                           
+     wait                  = optional(bool)                                           
+     extra_values          = optional(string)                                         
+     name                  = optional(string)                                         
+     namespace             = optional(string)                                         
+     chart                 = optional(string)                                         
+     repository            = optional(string)                                         
+     service_account_names = optional(object({                                        
+       controller = optional(string)                                                  
+       snapshot   = optional(string)                                                  
+     }))                                                                              
+     create_iam_resources_irsa = optional(bool)                                       
+     create_storage_class      = optional(bool)                                       
+     storage_class_name        = optional(string)                                     
+     is_default_class          = optional(bool)                                       
+     enabled                   = optional(bool)                                       
+     chart_version             = optional(string)                                     
+     version                   = optional(string)                                     
+     iam_policy_override       = optional(string)                                     
+     default_network_policy    = optional(bool)                                       
+   })                                                                                 
+   default = {}                                                                       
+ }                        
```
It remplace the variable with `null` and the merge use `null` as a value as the merge is done from local to the vars in that order.

I don't really need to enforce validation I'm trying to use this feature only for documentation but maybe there is some other way

**Comment by fayak** [comment]
One solution I found was to use YAML as source for variables.
For example, for a list of servers:

servers.yml
```
server: &default
  cpu: 8
  memory: 8192
  ips: []

servers:
  server1:
    <<: *default
    cpu: 2
    ips: ["10.2.8.124"]
  server2:
    <<: *default
    cpu: 4
```

main.tf
```
locals {
    servers = yamldecode(file("servers.yml")).servers
}
```

Each server in variable servers will contain the 3 attributes defined, either the default or the overridden ones. Again, maybe not as good as implemented directly in terraform I reckon, but at least it works for us

**Comment by thiagolsfortunato** [comment]
I'm try to use `optionals()` and `defaults()` on object type but return error: 

```bash
$ terraform -v
Terraform v0.14.6
+ provider registry.terraform.io/hashicorp/aws v3.29.0
```

variables.tf
```terraform
variable "cert" {
  description = "ACM Certificate Attributes"
  type = object({
    domain_name               = string
    subject_alternative_names = list(string)
    validation_method         = string
    tags                      = map(string)

    options = optional(object({
      certificate_transparency_logging_preference = optional(string)
    }))
  })
}
```

main.tf
```terraform
locals {
  cert = defaults(var.cert, {
    options = {
      certificate_transparency_logging_preference = ""
    }
  })
}

resource "aws_acm_certificate" "cert" {
  domain_name       = local.cert.domain_name
  validation_method = local.cert.validation_method

  subject_alternative_names = local.cert.subject_alternative_names

  dynamic "options" {
    for_each = local.cert.options.certificate_transparency_logging_preference == "" ? [] : [var.cert.options]

    content {
      certificate_transparency_logging_preference = options.value.certificate_transparency_logging_preference
    }
  }

  tags = local.cert.tags
}
```

module.tf
```terraform
module "my-cert" {
  source = "../../../../tf-module-certificate-manager"

  cert = {
    domain_name               = "example.com"
    validation_method         = "DNS"
    subject_alternative_names = ["*.example.com"]

    tags = {
      Name = "example.com"
    }

    # options = {
    #   certificate_transparency_logging_preference = "ENABLED"
    # }
  }
}
```

This configuration return the follow error:

```bash
Error: Error in the function call

  on ../../../../tf-module-certificate-manager/main.tf line 37, in locals:
  37:   cert = defaults(var.cert, {
  38:     options = {
  39:       certificate_transparency_logging_preference = ""
  40:     }
  41:   })
    |----------------
    | var.cert is object with 5 attributes

Call to function "defaults" failed: panic in function implementation:
interface conversion: interface {} is nil, not map[string]interface {}
goroutine 10177 [running]:
runtime/debug.Stack(0xc004349268, 0x24f9dc0, 0xc004358a80)
        /usr/local/go/src/runtime/debug/stack.go:24 +0x9f
github.com/zclconf/go-cty/cty/function.errorForPanic(...)
        /go/pkg/mod/github.com/zclconf/go-cty@v1.7.1/cty/function/error.go:44
github.com/zclconf/go-cty/cty/function.Function.Call.func1(0xc00434a2f0,
0xc00434a310)
        /go/pkg/mod/github.com/zclconf/go-cty@v1.7.1/cty/function/function.go:291
+0x95
panic(0x24f9dc0, 0xc004358a80)
        /usr/local/go/src/runtime/panic.go:969 +0x1b9
github.com/zclconf/go-cty/cty.Value.GetAttr(0x2ce7960, 0xc001852b50, 0x0, 0x0,
0xc00098f5f0, 0x2b, 0x0, 0x0, 0x2ce7860, 0xc0000503d9)
        /go/pkg/mod/github.com/zclconf/go-cty@v1.7.1/cty/value_ops.go:755 +0x40b
github.com/hashicorp/terraform/lang/funcs.defaultsApply(0x2ce7960,
0xc001852b50, 0x0, 0x0, 0x2ce7960, 0xc0025df9e0, 0x24ba300, 0xc004358750,
0x24ba300, 0xc004358750, ...)
        /home/circleci/project/project/lang/funcs/defaults.go:102 +0x385
github.com/hashicorp/terraform/lang/funcs.defaultsApply(0x2ce7960,
0xc0025df910, 0x24ba300, 0xc004358360, 0x2ce7960, 0xc0025df9f0, 0x24ba300,
0xc0043587b0, 0x0, 0xc00434a0d8, ...)
        /home/circleci/project/project/lang/funcs/defaults.go:107 +0x45c
github.com/hashicorp/terraform/lang/funcs.glob..func29(0xc004304a40, 0x2, 0x2,
0x2ce7960, 0xc0025df910, 0xc0025dfb90, 0x24ba300, 0xc0043589c0, 0xc004358a20,
0xc00434a190, ...)
        /home/circleci/project/project/lang/funcs/defaults.go:65 +0xb3
github.com/zclconf/go-cty/cty/function.Function.Call(0xc00021a990,
0xc004304a40, 0x2, 0x2, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0)
        /go/pkg/mod/github.com/zclconf/go-cty@v1.7.1/cty/function/function.go:295
+0x51a
github.com/hashicorp/hcl/v2/hclsyntax.(*FunctionCallExpr).Value(0xc00104b0e0,
0xc0019ff0a0, 0x0, 0xc00434b800, 0x1, 0x1, 0x0, 0x0, 0x0)
        /go/pkg/mod/github.com/hashicorp/hcl/v2@v2.8.2/hclsyntax/expression.go:442
+0x10c5
github.com/hashicorp/terraform/lang.(*Scope).EvalExpr(0xc003f4efa0, 0x2ce63e0,
0xc00104b0e0, 0x2ce78a0, 0x3dfc420, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
        /home/circleci/project/project/lang/eval.go:171 +0x1b7
github.com/hashicorp/terraform/terraform.(*BuiltinEvalContext).EvaluateExpr(0xc003ea9ee0,
0x2ce63e0, 0xc00104b0e0, 0x2ce78a0, 0x3dfc420, 0x0, 0x0, 0x0, 0x200000003,
0xc0015b2000, ...)
        /home/circleci/project/project/terraform/eval_context_builtin.go:287 +0xbb
github.com/hashicorp/terraform/terraform.(*NodeLocal).Execute(0xc003eccab0,
0x2d23580, 0xc003ea9ee0, 0xc0035a0002, 0x2509020, 0x2690320)
        /home/circleci/project/project/terraform/node_local.go:156 +0x71d
github.com/hashicorp/terraform/terraform.(*ContextGraphWalker).Execute(0xc0031a2c30,
0x2d23580, 0xc003ea9ee0, 0x7fe67a2bd428, 0xc003eccab0, 0x0, 0x0, 0x0)
        /home/circleci/project/project/terraform/graph_walk_context.go:127 +0xbc
github.com/hashicorp/terraform/terraform.(*Graph).walk.func1(0x2690320,
0xc003eccab0, 0x0, 0x0, 0x0)
        /home/circleci/project/project/terraform/graph.go:59 +0x962
github.com/hashicorp/terraform/dag.(*Walker).walkVertex(0xc000d91140,
0x2690320, 0xc003eccab0, 0xc002cc9e40)
        /home/circleci/project/project/dag/walk.go:387 +0x375
created by github.com/hashicorp/terraform/dag.(*Walker).Update
        /home/circleci/project/project/dag/walk.go:309 +0x1246
.

Releasing state lock. This may take a few moments...
```

Someone could help me?

**Comment by ktmorgan** [comment]
> I'm try to use `optionals()` and `defaults()` on object type but return error:
> 
> ```shell
> $ terraform -v
> Terraform v0.14.6
> + provider registry.terraform.io/hashicorp/aws v3.29.0
> ```
> Someone could help me?

@thiagolsfortunato According to the doc, `defaults` is only available in 0.15+

https://www.terraform.io/docs/language/functions/defaults.html

**Comment by thiagolsfortunato** [comment]
> > I'm try to use `optionals()` and `defaults()` on object type but return error:
> > 
> > ```shell
> > $ terraform -v
> > Terraform v0.14.6
> > + provider registry.terraform.io/hashicorp/aws v3.29.0
> > ```
> > Someone could help me?
> 
> @thiagolsfortunato According to the doc, `defaults` is only available in 0.15+
> 
> https://www.terraform.io/docs/language/functions/defaults.html
> 

But it's available as experimental feature. 

https://www.terraform.io/docs/language/expressions/type-constraints.html#experimental-optional-object-type-attributes

````
terraform { 
   experiments = [module_variable_optional_attrs] 
}
````

**Comment by whiskeysierra** [comment]
The docs are not very clear on this, but I tried pretty much everything and I believe that you can't have optional structural types due to how the `defaults` functions treats those. It will basically try to get to the deepest level of your data structure before applying defaults.

What I ended up with was having very shallow objects and instead of defaults I'm using coalesce on each key/property individually. That works with optional objects/lists/etc at least.

**Comment by DenWin** [comment]
Does there exist any timeline as to when "module_variable_optional_attrs" - and defaults - won't be experimental anymore?

**Comment by thiagolsfortunato** [comment]
> Does there exist any timeline as to when "module_variable_optional_attrs" - and defaults - won't be experimental anymore?

when launching the new release 0.15.

**Comment by alex-feel** [comment]
How to use `defaults` function with a variable of type list of objects? I have a module, where (just a part of the module):

```hcl
variable "records" {
  type        = list(object({
    record_name = string
    type        = string
    name        = optional(string)
    value       = string
    priority    = optional(number)
    ttl         = optional(number)
    proxied     = optional(bool)
  }))
  description = "Zone's DNS records."
}
```

```hcl
resource "cloudflare_record" "this" {
  for_each = var.records != null ? {for record in var.records : record.record_name => record} : {}

  zone_id  = cloudflare_zone.this.id

  type     = each.value.type
  name     = each.value.name
  value    = each.value.value
  priority = each.value.priority
  ttl      = each.value.ttl
  proxied  = each.value.proxied
}
```

In my configuration I can have this:

```hcl
module "acme_com" {
  # Here are other attributes
  records       = [
    {
      record_name = "a"
      type        = "A"
      name        = "acme.com"
      value       = "157.131.143.13"
    },
    {
      record_name = "mx1"
      type        = "MX"
      value       = "aspmx.l.google.com"
      priority    = 1
      proxied     = false
    }
  ]
}
```

Different list items may have different optional attributes omitted. How can I set the values of the omitted attributes for each list item?

**Comment by DenWin** [comment]
> > Does there exist any timeline as to when "module_variable_optional_attrs" - and defaults - won't be experimental anymore?
> 
> @DenWin when launching the new release 0.15.

Sorry, but they still seem to be experimental.

**Comment by ericsonrumuy7** [comment]
> > > Does there exist any timeline as to when "module_variable_optional_attrs" - and defaults - won't be experimental anymore?
> > 
> > 
> > @DenWin when launching the new release 0.15.
> 
> Sorry, but they still seem to be experimental.

Any update on this ?
really need this feature :)

**Comment by SavvasRonTruex** [comment]
At this moment even if optional would be added that would be a huge step, there are other ways to handle default.

**Comment by tikicoder** [comment]
Tagging to follow issue

**Comment by vistpp** [comment]
Maybe someone could provide a likelihood this experiment will go to production? Really need it, and we are wondering about the risk using this experimental feature in production.

**Comment by grimm26** [comment]
I'm using it in production in a couple modules. You're get the annoying warning, but that's not too troublesome. They'll promote it I'm sure.

**Comment by laughtonsm** [comment]
Same  as @grimm26 here.  It's proving very useful, although the `try` function seems to not handle the optional properties too well.  Or it could just be me not using it correctly

**Comment by sidh** [comment]
Yep, `optional` is very useful. It simplifies our code a lot.

**Comment by speller** [comment]
The `optional` function works well. It also would be nice to have an additional parameter to define a default value instead of null.

**Comment by mponton-cn** [comment]
Adding my 👍🏻 on `optional` and the `module_variable_optional_attrs` flag. We've been using it for a while in our modules and I'm getting anxious to see this promoted to a real non-experimental feature. A rollback/removal of the feature would make me pretty sad...

**Comment by alex-feel** [comment]
> Same as @grimm26 here. It's proving very useful, although the `try` function seems to not handle the optional properties too well. Or it could just be me not using it correctly

[It works](https://github.com/alex-feel/terraform-cloudflare-zone/blob/main/variables.tf#L588) well for me.

**Comment by Roxyrob** [comment]
Optional is a great feature that simplify a lot the code but as others already said, alongside `default null` behavior, the ability to define a `custom default`:
```
argumentN optional(bool, true)
```
will really complete "optional" feature greatly improving DRY and allowing a single point for variables/arguments declaretion instead of duplication in variables + locals.

**Comment by md5** [comment]
> Optional is a great feature that simplify a lot the code but as others already said, alongside `default null` behavior, the ability to define a `custom default`:
> 
> ```
> argumentN optional(bool, true)
> ```
> 
> will really complete "optional" feature greatly improving DRY and allowing a single point for variables/arguments declaretion instead of duplication in variables + locals.

It's more than a matter of duplication. Having the default declared inline as part of the variable declaration allows documentation to be generated automatically without the risk of the default and locals getting out of sync.

**Comment by lupass93** [comment]
indispensable and fantastic functionality! as they have implemented it now! I hope that from the next version the warning of the experimental functionality will disappear and become stable! I use it in every module! I can’t help it! totally in agreement with @prehor

**Comment by marceloboeira** [comment]
> I think the defaults() stuff is overly cumbersome for most cases and unnecessarily separates default values from the rest of the schema definition, TBH), but frankly I'd be perfectly happy if the current behavior was just made non-experimental as-is.

I guess they are different features at the end, I see use for both, but totally agree that trying to use `defaults` to cover the `optional` behavior is just... ugly.

**Comment by amarkevich** [comment]
@hashicorp Take 1.5+k votes into account during next feature planning plz

**Comment by jonwtech** [comment]
To be honest I use `module_variable_optional_attrs` so extensively now that if it's removed I'll just have to stay on my current version of TF forever :) - it's been a godsend. Agree that an optional second parameter for a default value would make it perfect. The current `defaults` function doesn't behave intuitively for me when dealing with maps and lists.

**Comment by code-haven** [comment]
Using `optional` + `default` has great potential for module developers to express complex & nested inputs like many yaml/json based templating engines(helm etc).

**Comment by Vinayaks439** [comment]
Hey Guys! I'm facing similar issue when creating optional variables inside a list of objects. Is this Issue related to that or did I missing something in the comments for workarounds. 

`terraform version: 0.13.7`

`variable "foo" {
    type = list(object({
        bar= optional(bool),
        hello= optional(bool),
        bye= optional(bool),
        name = string,
        description = string,
        birthyear = optional(number),
        age = optional(number),
        text = string,
        records= list(object({
            name = string
            description = string
            alias = string
        }))
    }))
}`

Issue: `Keyword "optional" is not a valid type constructor.`

**Comment by jonwtech** [comment]
> Hey Guys! I'm facing similar issue when creating optional variables inside a list of objects. Is this Issue related to that or did I missing something in the comments for workarounds.
> 
> `terraform version: 0.13.7`
> 
> `variable "foo" { type = list(object({ bar= optional(bool), hello= optional(bool), bye= optional(bool), name = string, description = string, birthyear = optional(number), age = optional(number), text = string, records= list(object({ name = string description = string alias = string })) })) }`
> 
> Issue: `Keyword "optional" is not a valid type constructor.`

You need to be on at least v0.14

**Comment by gmauleon** [comment]
Another 2022 bump on this, would definitely need it in my map of object setup where I don't want advanced options to be mandatory.

**Comment by prowlaiii** [comment]
I find the separate declaration of the types and the defaults a bit cumbersome and the "optional" keyword superfluous.

I think this has already been suggested, but the syntax for declaring the attributes of a variable are already defined, so it would be consistent to just make them recursive, eg.

`
variable "simple1" {
  default = "blah"
  description = "A simple variable"
  type = string
}
variable "complex1" {
  default = null
  description = "A complex variable"
  type = object({
    mystring1 = {
      default = "blah"
      description = "A string in an object variable (optional)"
      type = string
    }
    mystring2 = {
      description = "A string in an object variable (required)"
      sensitive = true
      type = string
    }
    mylist1 = {
      default = []
      description = "A list in an object variable (optional)"
      type = list(string)
    }
    mymap1 = {
      default = {}
      description = "A map in an object variable (optional)"
      type = map(string)
    }
    myobject1 = {
      default = null
      description = "An object within an object variable (optional)"
      type = object({
        ....
      )}
    }
  })
}
`

**Comment by ImIOImI** [comment]
I was surprised that when specifying an object as a type it filters out all inputs that aren't declared. I love all the ideas about specifying optional variables and stuff... but this has been open for 4 years. Maybe just don't kill things not declared in the type, then give us the awesome later?

tfvars:
```
test = {
  map0 = {
    required = "string"
  }
  map1 = {
    required = "string"
    optional = "string"
  }
}
```

variable:
```
variable "test" {
  type = map(
    object(
      {
        required = string
      }
    )
  )
}
```

outputs:
```
test = tomap({
  "map0" = {
    "required" = "string"
  }
  "map1" = {
    "required" = "string"
  }
})
```

**Comment by pascal-hofmann** [comment]
> I was surprised that when specifying an object as a type it filters out all inputs that aren't declared. I love all the ideas about specifying optional variables and stuff... but this has been open for 4 years. Maybe just don't kill things not declared in the type, then give us the awesome later?
```

I prefer to have them declared explicitly. This way you can look at the type declaration and see which fields are supported.

**Comment by dylanturn** [comment]
> I prefer to have them declared explicitly. This way you can look at the type declaration and see which fields are supported.

It also means that the language server (`terraform-ls`/`terraform-lsp`) can introspect the modules and provide much more accurate auto-complete suggestions...

**Comment by ImIOImI** [comment]
> I prefer to have them declared explicitly. This way you can look at the type declaration and see which fields are supported.

Me too... but the five stages of grief are "denial, anger, bargaining, depression and acceptance" and right now I'm at the bargaining stage. If I can't have something perfect, can I at least have something less nice?

**Comment by mponton-cn** [comment]
I'm at the bargaining phase myself. Simply getting rid of the "experimental" warning would please me.

**Comment by tikicoder** [comment]
For me this has been experimental for how long, I just would like to know what is the reason, what is causing issues, and is it really something that should be a blocker versus a hey every FYI this WILL break if you do this you know it will break.

**Comment by Nuru** [comment]
@apparentlymart Can we get an ETA (or declaration of lack of interest in an ETA) in taking [this feature](https://www.terraform.io/language/expressions/type-constraints#experimental-optional-object-type-attributes) out of `experimental` and into production? I would really like to be using it in a lot of places now. It has been experimental in 0.14, 0.15, 1.0, and 1.1. Can we get it into production at 1.2? CC: @alisdair

> Until the experiment is concluded, the behavior of this feature may see breaking changes even in minor releases. We recommend using this feature only in prerelease versions of modules as long as it remains experimental.

**Comment by apparentlymart** [comment]
Hi all,

Based on feedback about the existing experiment, we don't intend to graduate it to be a stable feature exactly as currently designed. Instead, we intend to make another design iteration which better integrates the idea of default values into the constraints system, and deals with some other concerns that folks raised in feedback on the existing proposal.

We don't intend to take the experiment away until there's a new experiment to replace it, but if you do wish to use it in production in spite of its experimental status I suggest keeping in mind that you will probably need to change your configuration at least a little in response to the changes for the second round of experiment.

The purpose of experiments is to gather feedback before finalizing a design, so in this case the experiment process worked as intended but since the feedback led to us needing another round of design we had to wait until there was time to spend on that design work, since the team was already wholly assigned to other work. We'll share more here when there's more to share.

**Comment by joe-a-t** [comment]
Flagging https://github.com/hashicorp/terraform/issues/30608 as a related issue that might merit consideration for future state consistency when refactoring of the optional arguments stuff here.

**Comment by eeckhardt-chwy** [comment]
It seems like long-term, to enforce types in a way that works for reasonable use cases, following in the pattern of TypeScript could be extremely useful. Let's say there are two properties that are mutually exclusive. Even with optional(), there's no way to express with the type system that either ONE or the OTHER must be present—that would have to be a validation error that occurs at runtime instead of giving an error at design time.

Here's a concrete example of a real use-case for creating AWS CloudWatch dashboard widget JSON: properties `query` and `metrics` are mutually exclusive, but one is required. Using `optional()` might look like this:

```
variable widget_definition {
  type = object({
    metrics = optional(list(list(string)))
    query = optional(string)
    ... other attributes here
  })
  description = "An object similar in structure to the JSON for CloudWatch dashboard widgets, but with some default values and simplifications/helpers to make the input simpler."
  validation {
    condition = count([ for key in var.widget_definition: key if contains(["metrics", "query"], key) ]) == 1
    error_message = "Must provide exactly one of properties 'metrics' and 'query' in variable widget_definition."
  }
```

But TypeScript handles this with type unions. That might look something like this:

```
  type = object({
     ... other attributes here
  }) & (
    { metrics = list(list(string) }
    | { query = string }
  )
```

Using something like this would avoid the need for validation to enforce the mutual exclusivity, allowing the type system itself to represent the strict expected types and the either/or nature of both `metrics` and `query`.

**Comment by fabiendelpierre** [comment]
Looks like this is happening with TF v1.3.0.

https://github.com/hashicorp/terraform/releases/tag/v1.3.0-alpha20220608

**Comment by korinne** [comment]
@fabiendelpierre You beat me to it :) Hi all, we'd love your feedback on the latest v1.3 alpha, which includes the ability to mark object type attributes as optional and set default values. You can learn more/provide feedback[ here](https://discuss.hashicorp.com/t/request-for-feedback-optional-object-type-attributes-with-defaults-in-v1-3-alpha/40550). Thank you in advance!

**Comment by prowlaiii** [comment]
> variable "simple1" {
>   default     = "blah"
>   description = "A simple variable"
>   type        = string
> }
> variable "complex1" {
>   default = null
>   description = "A complex variable"
>   type = object({
>     mystring1 = {
>       default = "blah"
>       description = "A string in an object variable (optional)"
>       type = string
>     }
>     mystring2 = {
>       description = "A string in an object variable (required)"
>       sensitive = true
>       type = string
>     }
>     mylist1 = {
>       default = []
>       description = "A list in an object variable (optional)"
>       type = list(string)
>     }
>     mymap1 = {
>       default = {}
>       description = "A map in an object variable (optional)"
>       type = map(string)
>     }
>     myobject1 = {
>       default = null
>       description = "An object within an object variable (optional)"
>       type = object({
>         ....
>       )}
>     }
>     mystring3 = { description = "Another string in an object variable (required)", type = string }
>   })
> }

Further to my previous post, I would suggest that (though the v1.3.0 adding of the 2nd default field in the "optional()" function is an improvement) the "optional()" keyword/function itself is really a bit of a red-herring and simply expanding the variable structure to allow nesting would be a more natural and syntactically consistent approach.

A simple variable declaration:
`variable "name" {
  type     = string
  default = "blah"
}`

A complex variable declaration:
`variable "settings" {
  type = object({
    name             = string       # shorthand for "{ type = string }"
    description    = { type = string, default = "The name" }
    how_many     = { type = number, default = 1, description = "Count of items" }
    list_of_items = { type = list, default = [] }
  })
}`

Just as with normal variables, the act of declaring a default makes it optional; there is no need for an "optional()" keyword/function and conversely not specifying a "default" makes it a required item.

In fact, the "object()" declaration could become redundant too, as the fact that there is a mix of types would implicitly make that so.
`variable "settings" {
  type = {
    name             = string       # shorthand for "{ type = string }"
    description    = { type = string, default = "The name" }
    how_many     = { type = number, default = 1, description = "Count of items" }
    list_of_items = { type = list, default = [] }
  }
}`

Extrapolating that further, the "type =" lines could be recursive and thus support sub-objects .

**Comment by atavakoliyext** [comment]
@prowlaiii is what you're describing similar to the feature request in #22449?

**Comment by prowlaiii** [comment]
> @prowlaiii is what you're describing similar to the feature request in #22449?

I think it is indeed; I hadn't seen it, but I knew I couldn't have been the first person on the planet to think of it!

(I'm also suggesting the "description =" item and removing the need for "object()" there.)

FYI, I also raised this in the v1.3 feedback and have now included a link to yours there.
[https://discuss.hashicorp.com/t/request-for-feedback-optional-object-type-attributes-with-defaults-in-v1-3-alpha/40550/47](url)

**Comment by dylanturn** [comment]
> A complex variable declaration:
> 
> ```
> variable "settings" {
>   type = object({
>     name          = string       # shorthand for "{ type = string }"
>     description   = { type = string, default = "The name" }
>     how_many      = { type = number, default = 1, description = "Count of items" }
>     list_of_items = { type = list, default = [] }
>   })
> }
> ```

If we're going to take it that far then why not just go all the way and create a new type that lets you simply specify an openapi schema?

The openapi shema variable type could be defined inline, sourced from a file in the same directory, or maybe even from a remote source (like we source modules today). That capability would make it easy to create, document, and reuse complex data classes in our input and output blocks.

**Comment by apparentlymart** [comment]
Hi all,

An evolution of the optional attributes experiment, with built-in support for default values and no separate `defaults` function, is going to graduate to stable in the forthcoming v1.3.0 release.

For that reason, I'm going to close this issue to represent that the base use-case is met. I see that there is some discussion here about further additions to the model for object attributes, but I think it'll be better to discuss those in more specific issues once the initial features are included in a stable release and we can learn from experiences of using the first iteration of this functionality in a broad set of Terraform modules with differing design needs.

Thanks for the great discussion here!
